Add new error type for failed resource reads

stlaz opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:
This PR implements an error that allows handling specific to errors that occur when decoding/transforming data from the database during read operations.

#### Which issue(s) this PR fixes:
Related to https://github.com/kubernetes/enhancements/issues/3926

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
A new error type was added to distinguish errors that occur when decoding/transforming resource data from the backend database. This error appears as a `StatusError` having `Reason` set to "StorageReadError".

The HTTP status code for the error remains at 500, as it would for any other internal server error.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
- [KEP]: https://github.com/kubernetes/enhancements/issues/3926
```

An example of such an error might look something like this:
```go
&errors.StatusError{
  ErrStatus: v1.Status{
    TypeMeta: v1.TypeMeta{
      Kind:       "",
      APIVersion: "",
    },
    ListMeta: v1.ListMeta{
      SelfLink:           "",
      ResourceVersion:    "",
      Continue:           "",
      RemainingItemCount: (*int64)(nil),
    },
    Status:  "Failure",
    Message: "failed to read one or more secrets from the storage; failed keys: [/ac770294-4d69-4a8e-a601-4ab7435b5d05/registry/secrets/secret-encryption-test/pre-fail-a /ac770294-4d69-4a8e-a601-4ab7435b5d05/registry/secrets/secret-encryption-test/pre-fail-b]",
    Reason:  "StorageReadError",
    Details: &v1.StatusDetails{
      Name:   "/secrets",
      Group:  "",
      Kind:   "secrets",
      UID:    "",
      Causes: []v1.StatusCause{
        v1.StatusCause{
          Type:    "UnexpectedServerResponse",
          Message: "cipher: message authentication failed",
          Field:   "/ac770294-4d69-4a8e-a601-4ab7435b5d05/registry/secrets/secret-encryption-test/pre-fail-a",
        },
        v1.StatusCause{
          Type:    "UnexpectedServerResponse",
          Message: "cipher: message authentication failed",
          Field:   "/ac770294-4d69-4a8e-a601-4ab7435b5d05/registry/secrets/secret-encryption-test/pre-fail-b",
        },
      },
      RetryAfterSeconds: 0,
    },
    Code: 500,
  },
}
```

/assign @deads2k 
/assign @dgrisonnet

stlaz committed:
Add StoreRead error and use it in the etcd store

The StoreRead error is returned from the etcd store when a
resource fails to be decoded or transformed. The former typically
occurs because of stored data corruption whereas the latter is often
a result of broken encryption configuration.

stlaz committed:
integration: remove APIServer config after its shutdown

This is to prevent unnecessary errors popping out from any dynamic
config reader (like EncryptionConfig, JWTAuthenticationrs etc.).

dgrisonnet commented on staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go:
```diff
@@ -902,6 +902,22 @@ const (
+	//      - (optional):
+	//        - "type" CauseType - CauseTypeUnexpectedServerResponse
+	//        - "message" string - the error message from the store backend
+	//        - "field" string - the full path with the key of the resource that failed reading
```
Maybe we could rename this to `key` or `resourceKey/objectKey` to make it clearer to the users what that corresponds to

stlaz replied:
We cannot rename any of these fields. The structure of a `StatusError` is provided - https://github.com/kubernetes/kubernetes/blob/29f1cdbf995a6c6877d5694f90d73a209a42b148/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go#L746-L747

We can only choose how to interpret each of these fields.

dgrisonnet replied:
Maybe we could use StatusDetails then? https://github.com/kubernetes/kubernetes/blob/29f1cdbf995a6c6877d5694f90d73a209a42b148/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go#L785-L797

dgrisonnet commented on staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go:
```diff
@@ -902,6 +902,22 @@ const (
+	//        - "type" CauseType - CauseTypeUnexpectedServerResponse
+	//        - "message" string - the error message from the store backend
+	//        - "field" string - the full path with the key of the resource that failed reading
+	//
```
What would you think about also adding fields that could easily be copied to kubectl? Like for example:
- Resource = GVR
- ResourceName
That could be easier to use with kubectl. We could even merge both fields to make it simpler.

But at the end of the day it depends on which tool we want to use to delete resources. The key would be nice to have when using `etcdctl` and the full resource name would be more appropriate to `kubectl`

stlaz replied:
I wonder if we can use the `TypeMeta` and `ListMeta` from the above-linked `StatusError`. This all happens pretty low in the stack and the data we have is quite limited.

dgrisonnet replied:
Maybe we could use StatusDetails no? https://github.com/kubernetes/kubernetes/blob/29f1cdbf995a6c6877d5694f90d73a209a42b148/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go#L785-L797

stlaz replied:
Those are actually used here: https://github.com/kubernetes/kubernetes/pull/124674/files#diff-c6a5b875364aa1a724d30dde8081e897a5af7f24b09b48dbaf7cbf88455b5516R395-R398
though I'm not sure about the `Kind: qualifiedResource.Resource,` line

dgrisonnet replied:
Oh I see, I missed it when I had a second look. I was mostly going with the example you shared above:
```
    Details: &v1.StatusDetails{
      Name:   "/secrets",
      Group:  "",
      Kind:   "secrets",
      UID:    "",
```

It looks like we might want to fix the name value so that the actual name of the secret is exposed?

stlaz replied:
Hmm, that's true. This error comes from listing, I think perhaps in that case the `Name` field should be omitted.

A get error currently looks like this:
```go
...
Details: &v1.StatusDetails{
  Name:   "/secrets/secret-encryption-test/pre-fail-c",
  Group:  "",
  Kind:   "secrets",
   UID:    "",
   Causes: []v1.StatusCause{
        v1.StatusCause{
           Type:    "UnexpectedServerResponse",
           Message: "cipher: message authentication failed",
           Field:   "/143d59bb-763f-4360-a86c-e7400375bfd6/registry/secrets/secret-encryption-test/pre-fail-c",
         },
    },
}
...
```
I suppose we could use the leaf element of the key instead of the full resource prefix in the storage. On the other hand, at least in the etcd store implementation, there is no concept of a "name" as that's not how the data is stored.

However, I've only literally just now noticed these [storage->API error transformators](https://github.com/kubernetes/kubernetes/blob/f307b0ae9d01c69256c2f72b7f50fd6022fc1941/staging/src/k8s.io/apiserver/pkg/storage/errors/storage.go#L40-L68) which I didn't see while writing the original enhancement.

I'll check if we could use these in this case to fill the status details with more relevant info.

dgrisonnet replied:
Maybe we could stick to the key in storage since it provides all the information that are needed to find the resource and just make it so that the utility to delete the resource can convert `/secrets/secret-encryption-test/pre-fail-c` into `/namespaces/secret-encryption-test/secrets/pre-fail-c`?

Or we could convert to an API path directly in the status so that it is also human readable

stlaz replied:
I agree that sticking to the storage key and later converting it to the proper resource path makes most sense to me.

I'd rather use the raw data we have directly from the etcd client and do any modifications necessary for any further work with them in tooling later.

cici37 commented:
/triage accepted

stlaz committed:
create a storage error for resource reads

Errors from the database backend are usually not handled directly
as API errors, but use a lower layer error library that's later
transformed to an API error by the layer that performed the higher-level
database request and has more context about the resource being
requested.

stlaz commented:
The latest patch adds an underlying error to the store errors first before this is translated into an API error to match the behavior of other store errors.

I'll add the errors format in an attachment to this comment.
[errors.txt](https://github.com/user-attachments/files/15786928/errors.txt)

dgrisonnet commented on staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go:
```diff
@@ -762,7 +763,8 @@ func (s *store) GetList(ctx context.Context, key string, opts storage.ListOption
 
 			data, _, err := s.transformer.TransformFromStorage(ctx, kv.Value, authenticatedDataString(kv.Key))
```
Should we consider wrapping TransformFromStorage errors?

I think that the current error that we expose in the status makes it hard to understand what's happening:
```
Type:    "UnexpectedServerResponse",
Message: "cipher: message authentication failed",
```

stlaz replied:
This is already part of the `StatusDetails.Causes`, right?

dgrisonnet replied:
Yes but I am wondering if from a user perspective it would be easy to understand that `cipher: message authentication failed` means that the server couldn't decrypt the secret because the key couldn't be retrieved. But maybe having it said explicitly in the Status.Message is clear enough?

dgrisonnet replied:
Nevermind, thinking about it again, I've been too focused on the reporting aspect, which should've already been covered by the KMS v2 effort. As you mentioned in another comment, this kep is mainly about unblocking scenario where the data can't be deleted and recreated. So reporting the message as is is fine.

dgrisonnet commented on staging/src/k8s.io/apiserver/pkg/storage/errors.go:
```diff
@@ -46,6 +48,7 @@ var errCodeToMessage = map[int]string{
 	ErrCodeInvalidObj:               "invalid object",
 	ErrCodeUnreachable:              "server unreachable",
 	ErrCodeTimeout:                  "request timeout",
+	ErrCodeCorruptedData:            "corrupted data",
```
why not undecryptable data? is this covering another scenario aswell?

stlaz replied:
As requested in the KEP, we're also handling corrupted data with this error.

dgrisonnet commented on staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go:
```diff
@@ -776,7 +778,8 @@ func (s *store) GetList(ctx context.Context, key string, opts storage.ListOption
 			if err != nil {
 				recordDecodeError(s.groupResourceString, string(kv.Key))
-				return err
+				failedKeys[string(kv.Key)] = err
```
In an ideal world undecryptable data can be recovered if the key becomes accessible again whereas corrupted data is completely lost. The only way to fix any residual problem is to delete the data completely and recreate it manually. Wouldn't it make sense to handle both of these scenarios separately to highlight that difference to the users?

stlaz replied:
> In an ideal world undecryptable data can be recovered if the key becomes accessible again whereas corrupted data is completely lost.

Scenario: KAS and etcd communicate over a network with aging hardware and a bit gets flipped every now and then. This way your data may be temporarily corrupt.
This enhancement is mostly meant to unblock execution by a deletion of undecryptable/corrupt resources that can be automatically recovered (e.g. recreated by a controller), or are no longer needed - see the [user stories in the enhancement](https://github.com/kubernetes/enhancements/blob/master/keps/sig-auth/3926-handling-undecryptable-resources/README.md#user-stories-optional).

The transformer in this part of the codebase is a generic construct and may not be exclusive to encryption - even though that's probably the only thing using it today. A more generic error seems appropriate.

We could probably special-case just the encryption error in the specific transformer and have a special API error just for that, but I can't say that I'm a big fan of such an approach.

dgrisonnet replied:
> The transformer in this part of the codebase is a generic construct and may not be exclusive to encryption - even though that's probably the only thing using it today. A more generic error seems appropriate.

That's a good point. I was too focused on the encryption use-case which made it seems like we were handling two completely different errors in the same way. But I think it is fine to have both of them reported the same way and then users can rely on the Status Causes to figure out what is exactly happening.

tkashem commented on staging/src/k8s.io/apimachinery/pkg/api/errors/errors.go:
```diff
@@ -375,6 +376,33 @@ func NewServerTimeout(qualifiedResource schema.GroupResource, operation string,
+		causes = append(causes, metav1.StatusCause{
+			Type:    metav1.CauseTypeUnexpectedServerResponse,
+			Field:   failedKey,
+			Message: err.Error(),
```
question: including the error message in the response, it might have a security concern?

stlaz replied:
I'd like to think not but I'll try to have a look

tkashem commented on staging/src/k8s.io/apiserver/pkg/storage/errors.go:
```diff
@@ -96,14 +99,40 @@ func NewInvalidObjError(key, msg string) *StorageError {
 type StorageError struct {
 	Code               int
 	Key                string
+	KeyErrors          map[string]error
```
An alternate that extends `StorageError` with an inner error:
```
--- a/staging/src/k8s.io/apiserver/pkg/storage/errors.go
+++ b/staging/src/k8s.io/apiserver/pkg/storage/errors.go
@@ -99,19 +99,20 @@ func NewInvalidObjError(key, msg string) *StorageError {
        }
 }
 
-func NewCorruptedDataError(errToKey map[string]error) error {
+func NewCorruptedDataError(key string, innerErr error) error {
        return &StorageError{
-               Code:      ErrCodeCorruptedData,
-               KeyErrors: errToKey,
+               Code:     ErrCodeCorruptedData,
+               Key:      key,
+               InnerErr: innerErr,
        }
 }
 
 type StorageError struct {
        Code               int
        Key                string
-       KeyErrors          map[string]error
        ResourceVersion    int64
        AdditionalErrorMsg string
+       InnerErr           error
 }
  
+func (e *StorageError) Unwrap() error { return e.InnerErr }
 ```

and inside etcd3/store.go, instead of using `failedKeys := make(map[string]error)`, we can use `errs := make([]error, 0)`, and then use aggregate error package https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/util/errors/errors.go#L46.

stlaz replied:
I like the idea that we hold on to the knowledge of keys for a little bit longer. I think we could implement unwrap to the aggregate error even with the current struct. Only I'm not sure if the aggregated error provides you with any more specific information, which I think is the usual reason why you would call it.

stlaz commented on staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go:
```diff
@@ -762,7 +763,8 @@ func (s *store) GetList(ctx context.Context, key string, opts storage.ListOption
 			data, _, err := s.transformer.TransformFromStorage(ctx, kv.Value, authenticatedDataString(kv.Key))
 			if err != nil {
-				return storage.NewInternalErrorf("unable to transform key %q: %v", kv.Key, err)
+				failedKeys[string(kv.Key)] = err
```
Right, the new error should likely appear behind a feature gate, too. I'll think about which approach to take, I'd like the switch to be as simple as possible.

stlaz replied:
I localized the change to the store, but also added a feature-gated panic() in the error constructor. PTAL.

stlaz commented on staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go:
```diff
@@ -762,7 +763,8 @@ func (s *store) GetList(ctx context.Context, key string, opts storage.ListOption
 			if err != nil {
-				return storage.NewInternalErrorf("unable to transform key %q: %v", kv.Key, err)
+				failedKeys[string(kv.Key)] = err
+				continue
```
Hmm, is there a reason this check is not at the beginning of the for loop?

stlaz replied:
moved it further up the function

stlaz commented on staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go:
```diff
@@ -789,6 +792,9 @@ func (s *store) GetList(ctx context.Context, key string, opts storage.ListOption
 		}
+		if len(failedKeys) > 0 {
+			return storage.NewCorruptedDataError(failedKeys)
+		}
```
hmm, that's a good catch, I think this should be handled further below the all-wrapping for-loop?

stlaz committed:
add a FG for the StorageReadError, add the error on couple more places

stlaz committed:
etcd store: move list timeout handling before transformation

tkashem commented:
@stlaz I have been working on the other end of it, the deletion flow, i have a work in progress PR https://github.com/kubernetes/kubernetes/pull/127513. Both the listing of corrupt objects, and the unsafe deletion flow need to converge on the error pattern. I took a stab at the error pattern for  untransformable/undecodable at these commits: https://github.com/kubernetes/kubernetes/pull/127513/commits/a48aee4825e5f7cedf4616a479c3eb22ed60b681 
https://github.com/kubernetes/kubernetes/pull/127513/commits/9705f78c4e1baa7f4a28b92fb0428c56ca45deec

If you think we can use these patterns, then please feel free to adopt it in your PR, otherwise i will adopt the error pattern you suggest. Once https://github.com/kubernetes/kubernetes/pull/124674 merges, I will rebase my PR.

stlaz commented:
/close
in favour of https://github.com/kubernetes/kubernetes/pull/127513
